### ISSUE TITLE
Unsound corner case for calling methods on Object with NNBD

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#551]
Consider the following code (assuming NNBD is enabled):
```dart
class C {
  Never toString() => throw 'foo';
}
int f(C? c) => c.toString(); // (1)
main() {
  print(f(null).isEven); // Uh oh
}
```
As currently specified, at (1), the static type of the call `c.toString()` is `Never`, because it resolves to `C.toString()`, which has a more specific return type than `Object.toString()`.  So the upcast from `Never` to `int` is allowed.  But at runtime, `c` will be bound to `null`, so `c.toString()` will return `'null'`, which isn't an `int`.

Adding optional or named parameters to `toString` creates similar problems:
```dart
class C {
  String toString([int i]) => '$i';
}
String f(C? c, int i) => c.toString(i);
main() {
  print(f(null, 1)); // Uh oh
}
``` 

Similar problems apply to `hashCode`, `runtimetype`, and `noSuchMethod`.

**Comment by stereotype441** [comment]
I discussed this with @leafpetersen over IM (using a slightly different example) and he thought of two possible ways to close the soundness loophole.  Paraphrasing:

(1) We could add restrictions on how you can override `Object` methods, so the definitions of class `C` above would be illegal.  This has the disadvantage that it's a breaking change, since adding optional parameters to e.g. `toString` is allowed today.

(2) We could change the static typing rules so that if the type of `x` is nullable, then the type looked up for e.g. `x.toString` is the type from `Object`; if the user wants to get the more specific type, they need to either promote `x` to a non-nullable type, or use a `?.` to access the method.

**Comment by leafpetersen** [comment]
Essentially the issue is that we are allowing method calls on a union type in a limited context (`Object` methods), and we need to take that into account in the typing.  

A sound typing rule for  `x.f` where `x` has type `A | B` and `A.f` has type `T0 -> R0` and `B.f` has type `T1 -> R1` is to say that `x.f` has type `T0 & T1 -> R0 | R1`.  

In this case, `B` is always `Null`, and the method signatures on `Null` are the same as the ones on `Object`.  The override rules restrict all subclasses of `Object` such that `T1 <: T0` will always be true (unless it's a covariant override), so `T0 & T1` is safely approximated by `T1` (that is, the argument type, if any, on the `Object` method). And dually, the override rules restrict all subclasses of `Object` such that `R0 <: R1`, and so `R0 | R1` is safely approximated by `R1` (that is, the result type of the `Object` method).    For covariant overrides of `Object` methods, it's still fine to use the `Object` signature, since the call is checked, and the "true" argument type is `Object`.

So in other words we should use the `Object` signatures for the `Object` methods when we call them on nullable types.

**Comment by lrhn** [comment]
So the problem is that the static type of `C?.toString()` is the static type of `C.toString()`, not the union with the type of `null.toString()`. And as Leaf says, the argument types of the non-null type of the `C?` union must be subtypes of the `null` method's argument type (barring covariance, and that's unsafe anyway), and the return type must be a supertype.

I agree with the conclusion - use the `Object` signatures (aka. the `Null` signatures) when calling any method on a union type. 

(If we ever get general union types, we'll have to do something more general. I'd say use the signature of the most specific common super-interface of the types in that case, which would be `Object` here, except for `Never?`, where it's `Null`, giving the exact same result).

**Comment by eernstg** [comment]
We shouldn't actually need anything more general for unrestricted union types: The type of a member `m` for a receiver whose type is a union `A|B|C` must be the least upper bound of the types the member in each operand, `A.m | B.m | C.m`. This is just a tiny generalization of Leaf's rules.

But I think we should be a bit more strict with union types, because I'd prefer to maintain the property that the nominal nature of Dart types can be used to understand what's going on. So even if `A`, `B`, and `C` all have an `m`, I wouldn't allow an invocation like `abc.m()` where `abc` has type `A|B|C`, unless `m` in all three types is an override of a common declaration of `m` in some shared supertype of `A` and `B` and `C`.

But these things are guaranteed automatically with methods from `Object` and a union of the form `A | Null`, so we don't need to impose any extra restrictions in order to keep that door open.

**Comment by leafpetersen** [comment]
This is resolved.
